### PR TITLE
extend: Add HashiCorp Provider Design Principles page

### PR DESCRIPTION
**Pull request by bflad**
## Description

The Terraform Provider team often has to field questions about overall provider design and recommended patterns. This content (along with other bits of the Best Practices section) may warrant additional refactoring or inclusion on this page, but this is raised to start discussions. 😄

### CODE REVIEW

File: content/source/docs/extend/hashicorp-provider-design-principles.html.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+sidebar_current: "docs-extend-hashicorp-provider-design-principles"
+description: |-
+  Design principles for officially maintained Terraform Providers.
+---
+
+# HashiCorp Provider Design Principles
+
+Over time, the Terraform development community has gained large amounts of experience across a wide breadth of Application Programming Interfaces (APIs) supported by Terraform and its Software Development Kit (SDK). Recommended practices have emerged with designing the expected operator and maintainer experience. The principles below highlight the most prolific patterns when designing Terraform providers and associated resources, which guide HashiCorp design decisions.
+
+## Providers should focus on single API or SDK
```
Comment (aareet) [inline content/source/docs/extend/hashicorp-provider-design-principles.html.md]: missing "a" single API or SDK

Comment (bflad) [reply content/source/docs/extend/hashicorp-provider-design-principles.html.md]: Fixed 😄

**Comment by cgriggs01** [comment]
I love this! ❤️ Thanks all!

File: content/source/docs/extend/hashicorp-provider-design-principles.html.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+- Allowing maintainers to be experts in a single system or ecosystem.
+
+## Resources should represent a single API object/mapping/event
+
+A Terraform resource should be a declarative representation of single component, usually with Create, Read, Delete, and sometimes Update methods. In general, abstractions of multiple components in Terraform should be accomplished via [Terraform Modules](/docs/modules/), potentially hosted in the [Terraform Registry](https://registry.terraform.io/).
+
+The benefits of this practice include:
+
+- Giving operators the most flexibility for composing their environments as varying team structures may impose permission limitations or automation desires.
+- Assuring operators that there are no unexpected interations between multiple API components.
```
Comment (megan07) [inline content/source/docs/extend/hashicorp-provider-design-principles.html.md]: Is this supposed to be "interactions", I think it's missing the "c"

Comment (bflad) [reply content/source/docs/extend/hashicorp-provider-design-principles.html.md]: Fixed!

**Pull request merged**